src/call/mod.rs

```diff
@@ -108,6 +110,126 @@ impl RpcStatus {
     }
 }
 
+/// `MessageReader` is a zero-copy reader for the message payload.
+///
+/// To achieve zero-copy, use the BufRead API `fill_buf` and `consume`
+/// to operate the reader.
+pub struct MessageReader {
+    buf: *mut GrpcByteBuffer,
+    reader: GrpcByteBufferReader,
+    slice: Option<GrpcSlice>,
+    // Hack: apparently its lifetime depends on `self.slice`. However
+    // it's not going to be accessed by others directly, hence it's safe
+    // to mark it static here.
+    bytes: &'static [u8],
+    length: usize,
+}
+
+impl MessageReader {
+    /// Get the available bytes count of the reader.
+    #[inline]
+    pub fn pending_bytes_count(&self) -> usize {
+        self.length
+    }
+
+    pub fn read_directly(&mut self, buf: &mut [u8]) -> usize {
+        let amt = {
+            let bytes = self.fill_buf_directly();
+            if bytes.is_empty() {
+                return 0;
+            }
+            let amt = cmp::min(buf.len(), bytes.len());
+            if amt == 1 {
+                buf[0] = bytes[0];
+            } else {
+                buf[..amt].copy_from_slice(bytes);
+            }
+            amt
+        };
+        self.consume(amt);
+        amt
+    }
+
+    pub fn read_to_end_directly(&mut self, buf: &mut Vec<u8>) -> usize {
+        if self.length == 0 {
+            return 0;
+        }
+        buf.reserve(self.length);
+        let start = buf.len();
+        let mut len = start;
+        unsafe {
+            buf.set_len(start + self.length);
+        }
+        let ret = loop {
+            match self.read_directly(&mut buf[len..]) {
+                0 => break len - start,
+                n => len += n,
+            }
+        };
+        unsafe {
+            buf.set_len(len);
+        }
+        ret
+    }
+
+    pub fn fill_buf_directly(&mut self) -> &[u8] {
+        if self.bytes.is_empty() {
+            if self.length == 0 {
+                return &[];
+            }
+            let mut len = 0;
+            unsafe {
+                match self.slice {
+                    None => self.slice = Some(mem::zeroed()),
+                    Some(s) => grpc_sys::grpc_slice_unref(s),
+                }
+                let slice = self.slice.as_mut().unwrap();
+                let code = grpc_sys::grpc_byte_buffer_reader_next(&mut self.reader, slice);
+                debug_assert!(code != 0);
+                let ptr = grpc_sys::grpcwrap_slice_raw(slice, &mut len);
+                self.bytes = slice::from_raw_parts(ptr as _, len);
+            }
+        }
+        self.bytes
+    }
+}
+
+unsafe impl Sync for MessageReader {}
```

Comment:
em, strange to support Sync and Send for a Reader?

Can this Reader be used crossing threads?

Reply:
This is copied from #134 , I thought it's necessary for something else so I didn't delete it.